File: mobile/native/ios/MeteoRideShare/MeteoRideShareStore.swift
import Foundation

/// Drop box for routes that arrive from outside the app.
///
/// One producer writes here: the app itself, when another app opens a .gpx/.kml "in"
/// MeteoRide through `CFBundleDocumentTypes` → `SceneDelegate`. The web layer drains it
/// through `MeteoRideSharePlugin`.
///
/// There was a share extension once, and it was a second producer over the same App
/// Group folder — which is what the UUID token in `inboxFileName` is left over from. It
/// was removed because its share-sheet entry is indistinguishable from the document one
/// and takes the tap, and no app extension can open its containing app. Do not add one
/// back; docs/IOS.md step 5 has the whole account.
enum MeteoRideShareStore {

    /// Must match the App Group enabled on the App target.
    static let appGroupId = "group.cc.meteoride.app"

    private static let folderName = "IncomingRoutes"
    private static let allowedExtensions: Set<String> = ["gpx", "kml"]
    private static let maxBytes = 25 * 1024 * 1024
    private static let maxAge: TimeInterval = 24 * 60 * 60

    // Two shares can land in the same millisecond; this tells their file names apart
    // within one process. It cannot do so across processes: the app and the share
    // extension each start their own counter at zero over the same App Group folder,
    // so the same millisecond and the same count produce the same name and the atomic
    // write drops one of the two routes. `uniqueToken` closes that; the counter stays
    // because it is what keeps arrival order inside a burst from one process.
    private static var sequenceCounter = 0
    private static let sequenceLock = NSLock()
    private static func nextSequence() -> Int {
        sequenceLock.lock()
        defer { sequenceLock.unlock() }
        sequenceCounter += 1
        return sequenceCounter
    }

    /// Eight hex characters of a fresh UUID: unique wherever the counter is not.
    private static func uniqueToken() -> String {
        String(UUID().uuidString.prefix(8)).lowercased()
    }

    /// Shared folder, created on first use. Nil means the App Group is misconfigured.
    static var inboxURL: URL? {
        let fm = FileManager.default
        guard let container = fm.containerURL(forSecurityApplicationGroupIdentifier: appGroupId) else {
            NSLog("[MeteoRide] App Group \(appGroupId) unavailable — check entitlements")
            return nil
        }
        let dir = container.appendingPathComponent(folderName, isDirectory: true)
        if !fm.fileExists(atPath: dir.path) {
            try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    // MARK: - Writing

    /// Stores raw route bytes. Returns the stored file, or nil if it could not be saved.
    @discardableResult
    static func store(data: Data, suggestedName: String) -> URL? {
        guard let dir = inboxURL, !data.isEmpty, data.count <= maxBytes else { return nil }
        let millis = Int64(Date().timeIntervalSince1970 * 1000)
        let fileName = inboxFileName(millis: millis, sequence: nextSequence(), name: suggestedName)
        let dest = dir.appendingPathComponent(fileName)
        do {
            try data.write(to: dest, options: .atomic)
            return dest
        } catch {
            NSLog("[MeteoRide] could not store shared route: \(error.localizedDescription)")
            return nil
        }
    }

    /// Copies a file the system handed us (share sheet or "Open in…").
    @discardableResult
    static func ingest(fileURL: URL) -> Bool {
        // Only local files. `Data(contentsOf:)` happily accepts an https URL and
        // performs a blocking, untimed network download, and the share sheet hands
        // over a web URL whenever someone shares a link from Strava, Komoot or a
        // browser. Sharing a link is not a supported way to open a route.
        guard fileURL.isFileURL else {
            NSLog("[MeteoRide] ignoring shared web URL: \(fileURL.scheme ?? "?")")
            return false
        }

        // Files coming from other apps may be security-scoped.
        let scoped = fileURL.startAccessingSecurityScopedResource()
        defer { if scoped { fileURL.stopAccessingSecurityScopedResource() } }

        // `LSSupportsOpeningDocumentsInPlace` is false, so the system COPIES each opened
        // document into this app's own Documents/Inbox and hands over the copy. That copy
        // is ours, and nothing else ever deletes it — the 24-hour prune only sweeps the
        // App Group folder — so left alone every route ever opened stays in the sandbox
        // for good, at up to 25 MiB each.
        //
        // Discarded only on a DEFINITIVE answer: stored, or rejected for what the file
        // is. A read that fails is not a verdict — no space, an iCloud download that has
        // not finished, a bad moment — and deleting there would destroy the only copy of
        // a route the user did ask for, silently, with `ingest` returning false and
        // nothing announced. Those stay for `recoverSystemInbox` to retry.
        guard let data = readCapped(fileURL) else { return false }
        guard accepts(name: fileURL.lastPathComponent, data: data) else {
            discardSystemInboxCopy(fileURL)
            return false
        }
        guard store(data: data, suggestedName: fileURL.lastPathComponent) != nil else { return false }
        discardSystemInboxCopy(fileURL)
        return true
    }

    /// Ingests the copies sitting in `Documents/Inbox` that no run ever collected, and
    /// returns how many routes that recovered. Three things end up stranded there and
    /// nothing else would ever look: a read interrupted by the app being killed mid-
    /// ingest, a read that failed for a reason that may not repeat, and — once —
    /// everything that piled up before any of this deleted them at all.
    ///
    /// Anything past `maxAge` is dropped unread, which is the policy the App Group folder
    /// already has and what stops a file that fails every single time from being retried
    /// for ever. Note the age is the only bound here: the 25 MiB cap limits what is
    /// ACCEPTED, not the size of a copy the system has already written.
    @discardableResult
    static func recoverSystemInbox() -> Int {
        let fm = FileManager.default
        guard let inbox = systemInboxURL,
              let names = try? fm.contentsOfDirectory(atPath: inbox.path) else { return 0 }
        var recovered = 0
        for name in names.sorted() {
            let url = inbox.appendingPathComponent(name)
            let modified = (try? fm.attributesOfItem(atPath: url.path))?[.modificationDate] as? Date
            if let modified, Date().timeIntervalSince(modified) > maxAge {
                try? fm.removeItem(at: url)
                continue
            }
            if ingest(fileURL: url) { recovered += 1 }
        }
        return recovered
    }

    /// Where the system leaves the copies it makes of opened documents.
    private static var systemInboxURL: URL? {
        try? FileManager.default
            .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false)
            .appendingPathComponent("Inbox")
    }

    /// Deletes a file only when it sits in this app's own `Documents/Inbox`, which is
    /// where the system leaves the copies it makes. Guarded on that directory precisely
    /// so a URL another app still owns — in-place, or security-scoped — is never touched.
    private static func discardSystemInboxCopy(_ fileURL: URL) {
        guard let inbox = systemInboxURL?.resolvingSymlinksInPath().path else { return }
        guard fileURL.resolvingSymlinksInPath().path.hasPrefix(inbox + "/") else { return }
        try? FileManager.default.removeItem(at: fileURL)
    }

    /// Reads at most `maxBytes`, giving up as soon as the file turns out to be bigger.
    /// `Data(contentsOf:)` would load all of it first, and a route can be 25 MiB.
    static func readCapped(_ fileURL: URL) -> Data? {
        guard let stream = InputStream(url: fileURL) else { return nil }
        stream.open()
        defer { stream.close() }
        var data = Data()
        var buffer = [UInt8](repeating: 0, count: 64 * 1024)
        while true {
            let read = stream.read(&buffer, maxLength: buffer.count)
            if read < 0 { return nil }
            if read == 0 { return data }
            data.append(buffer, count: read)
            if data.count > maxBytes {
                NSLog("[MeteoRide] ignoring shared file larger than \(maxBytes) bytes")
                return nil
            }
        }
    }

    /// Apps share routes with all sorts of types and names, so judge by the file name
    /// or, failing that, by what is actually inside. Mirrors the Android side.
    static func accepts(name: String, data: Data) -> Bool {
        guard !data.isEmpty, data.count <= maxBytes else { return false }
        if allowedExtensions.contains((name as NSString).pathExtension.lowercased()) { return true }
        let head = data.prefix(2048)
        guard let start = String(data: head, encoding: .utf8)?.lowercased()
            ?? String(data: head, encoding: .isoLatin1)?.lowercased() else { return false }
        return start.contains("<gpx") || start.contains("<kml")
    }

    // MARK: - Reading

    static func pendingURLs() -> [URL] {
        guard let dir = inboxURL else { return [] }
        prune(in: dir)
        let items = (try? FileManager.default.contentsOfDirectory(atPath: dir.path)) ?? []
        // `.atomic` writes leave a `<name>.sb-XXXX` temp file in the same directory
        // for the instant of the rename; excluding anything that is not a finished
        // inbox name keeps that temp file from being read and deleted as if it were
        // a real, complete route.
        return items.filter(isInboxName).sorted().map { dir.appendingPathComponent($0) }
    }

    /// A route shared while the web layer never got to run would sit here forever.
    private static func prune(in dir: URL) {
        let fm = FileManager.default
        let items = (try? fm.contentsOfDirectory(atPath: dir.path)) ?? []
        for item in items {
            let url = dir.appendingPathComponent(item)
            guard let modified = (try? fm.attributesOfItem(atPath: url.path))?[.modificationDate] as? Date else { continue }
            if Date().timeIntervalSince(modified) > maxAge { try? fm.removeItem(at: url) }
        }
    }

    /// Returns the oldest pending route as text and removes it from the inbox.
    static func nextPending() -> (name: String, text: String)? {
        for url in pendingURLs() {
            defer { try? FileManager.default.removeItem(at: url) }
            guard let data = try? Data(contentsOf: url) else { continue }
            guard let text = decode(data), !text.isEmpty else { continue }
            return (displayName(of: url), text)
        }
        return nil
    }

    // MARK: - Helpers

    /// Timestamp+sequence+identity file name: unique across shares and across the two
    /// processes that write here, and sorts in arrival order. One character longer than
    /// the Android side's `%013d-%04d__name`, which needs no identity: only the one
    /// process ever writes that inbox.
    static func inboxFileName(millis: Int64, sequence: Int, name: String) -> String {
        // %d reads a 32-bit int and would truncate a 13-digit millisecond
        // timestamp; %lld/%ld match the actual 64-bit argument width.
        String(format: "%013lld-%04ld-%@__%@", millis, sequence % 10000, uniqueToken(), sanitize(name))
    }

    /// Matches a finished inbox entry — the current `<millis>-<seq>-<id>__name` scheme
    /// or either legacy one (`<millis>-<seq>__name`, and `<millis>__name` before that,
    /// with no zero padding and no sequence) —
    /// as opposed to an `.atomic` write's transient `<name>.sb-XXXX` sibling or
    /// unrelated junk (e.g. `.DS_Store`). Case-insensitive: `sanitize` only checks
    /// the extension case-insensitively and keeps whatever case the sender used, so
    /// a name can legitimately end in `.GPX` or `.KML`.
    private static let inboxNameRegex = try? NSRegularExpression(
        pattern: "^(?:\\d{13}-\\d{4}(?:-[0-9a-f]{8})?__.+\\.(?:gpx|kml)|\\d+__.+\\.(?:gpx|kml))$",
        options: [.caseInsensitive, .dotMatchesLineSeparators]
    )

    static func isInboxName(_ name: String) -> Bool {
        guard let regex = inboxNameRegex else { return false }
        let range = NSRange(name.startIndex..<name.endIndex, in: name)
        return regex.firstMatch(in: name, range: range) != nil
    }

    /// GPX is XML and normally UTF-8; some exporters still emit Latin-1.
    private static func decode(_ data: Data) -> String? {
        String(data: data, encoding: .utf8) ?? String(data: data, encoding: .isoLatin1)
    }

    /// Strips the timestamp prefix added by `store`.
    private static func displayName(of url: URL) -> String {
        let name = url.lastPathComponent
        guard let range = name.range(of: "__") else { return name }
        return String(name[range.upperBound...])
    }

    private static func sanitize(_ name: String) -> String {
        var base = name.isEmpty ? "route.gpx" : name
        base = base.replacingOccurrences(of: "/", with: "-")
        base = base.replacingOccurrences(of: ":", with: "-")
        // A newline or other control character in a shared name would otherwise sit
        // unescaped in the stored file name; `isInboxName`'s `.+` now spans line
        // separators too, but new writes should never produce one in the first place.
        base = String(String.UnicodeScalarView(base.unicodeScalars.map { $0.value < 0x20 ? "-" : $0 }))
        if !allowedExtensions.contains((base as NSString).pathExtension.lowercased()) {
            base += ".gpx"
        }
        return String(base.suffix(120))
    }
}

File: mobile/android/app/src/main/java/cc/meteoride/app/MeteoRideShareStore.java
package cc.meteoride.app;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.OpenableColumns;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Drop box for routes that arrive from outside the app: a GPX shared from another
 * app, or a .gpx/.kml opened from a file manager or a download.
 *
 * The web layer drains it through {@link MeteoRideSharePlugin}. Files are parked on
 * disk rather than kept in memory, because the content URI grant dies with the
 * intent and the process may be recreated before JavaScript asks for the route.
 */
final class MeteoRideShareStore {

    private static final String TAG = "MeteoRide";
    private static final String DIR = "incoming-routes";
    /** The intake ledger: routes received from an intent but not yet in the inbox. */
    private static final String LEDGER = "incoming-routes.pending";
    private static final Object LEDGER_LOCK = new Object();
    private static final List<String> ALLOWED_EXTENSIONS = Arrays.asList("gpx", "kml");
    private static final int MAX_BYTES = 25 * 1024 * 1024;
    private static final long MAX_AGE_MS = 24 * 60 * 60 * 1000L;
    // Two shares can land in the same millisecond; this tells their file names apart.
    private static final AtomicInteger SEQUENCE = new AtomicInteger();

    private MeteoRideShareStore() {}

    /** A route waiting to be handed to the web layer. */
    static final class Pending {
        final String name;
        final String text;

        Pending(String name, String text) {
            this.name = name;
            this.text = text;
        }
    }

    // ---------------------------------------------------------------- writing

    /** Copies whatever the intent pointed at. Returns false if it was not a route. */
    static boolean ingest(Context ctx, Uri uri) {
        if (uri == null) return false;

        String name = displayName(ctx, uri);
        byte[] data;
        try (InputStream in = ctx.getContentResolver().openInputStream(uri)) {
            if (in == null) return false;
            data = readAll(in);
        } catch (IOException | SecurityException e) {
            Log.w(TAG, "could not read shared route: " + e.getMessage());
            return false;
        }

        // Plenty of apps share a GPX as application/octet-stream with no usable file
        // name, so fall back to looking at the content itself.
        if (!hasAllowedExtension(name) && !looksLikeRoute(data)) {
            Log.w(TAG, "ignoring shared item, not a route: " + name);
            return false;
        }
        return store(ctx, data, name) != null;
    }

    static File store(Context ctx, byte[] data, String suggestedName) {
        if (data == null || data.length == 0) return null;
        // Timestamp+sequence prefix keeps arrival order and avoids collisions between
        // shares that land in the same millisecond.
        String fileName = inboxFileName(System.currentTimeMillis(), SEQUENCE.incrementAndGet(), suggestedName);
        File dest = new File(inbox(ctx), fileName);
        // Written elsewhere and moved in whole: ingestion runs on its own thread, and
        // next() would otherwise read, and delete, a file still being written. The
        // cache and files directories share a volume, so the rename is atomic.
        File part = new File(ctx.getCacheDir(), fileName + ".part");
        try (FileOutputStream out = new FileOutputStream(part)) {
            out.write(data);
        } catch (IOException e) {
            Log.w(TAG, "could not store shared route: " + e.getMessage());
            if (!part.delete()) Log.w(TAG, "could not delete " + part.getName());
            return null;
        }
        if (!part.renameTo(dest)) {
            Log.w(TAG, "could not move shared route into the inbox");
            if (!part.delete()) Log.w(TAG, "could not delete " + part.getName());
            return null;
        }
        return dest;
    }

    // ------------------------------------------------------------ intake ledger

    /*
     * Reading a shared file takes as long as the provider takes, and until the bytes
     * are in the inbox the route exists nowhere durable: the content URI dies with the
     * intent and the process can be killed in between. The activity used to mark the
     * intent handled before that read even started, so a death mid-read lost the route
     * outright — and with `savedInstanceState` and FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY
     * both refusing to read an intent twice, not even reopening from Recents got it
     * back. The user had to go and share it again.
     *
     * So the URI is written here, durably, before anything is read, and removed once
     * the attempt is over. What is left in this file at launch is exactly what was
     * received and never delivered, and {@link MainActivity} retries it. A delivery
     * that finished is not in the file, so it is never delivered twice.
     */

    /**
     * Records URIs as received. Call before reading them, not after. Returns false if
     * the record could not be written — a full disk, a read-only volume — and the
     * caller must then not treat the intent as handled: with nothing durable to come
     * back to, the only remaining chance is the intent itself.
     */
    static boolean rememberIntake(Context ctx, List<Uri> uris) {
        if (uris.isEmpty()) return true;
        List<String> lines = new ArrayList<>();
        for (Uri uri : uris) lines.add(uri.toString());
        synchronized (LEDGER_LOCK) {
            return writeLedger(ctx, ledgerWith(readLedger(ctx), lines));
        }
    }

    /** Drops one URI, whether its delivery succeeded or failed. One attempt, no loop. */
    static void forgetIntake(Context ctx, Uri uri) {
        synchronized (LEDGER_LOCK) {
            String before = readLedger(ctx);
            String after = ledgerWithout(before, uri.toString());
            if (!after.equals(before)) writeLedger(ctx, after);
        }
    }

    /**
     * Whether this URI is still waiting. This is the claim a delivery stakes: the entry
     * is only removed once an attempt is over, so a second task carrying the same URI
     * finds it gone and skips. Called from the ingest thread, which is the only thread
     * that removes, so the check and the delivery that follows cannot be raced by
     * another delivery.
     */
    static boolean isPendingIntake(Context ctx, Uri uri) {
        synchronized (LEDGER_LOCK) {
            return ledgerLines(readLedger(ctx)).contains(uri.toString());
        }
    }

    /** What was received and never delivered. Empty on a clean launch. */
    static List<Uri> pendingIntake(Context ctx) {
        String content;
        synchronized (LEDGER_LOCK) {
            content = readLedger(ctx);
        }
        List<Uri> uris = new ArrayList<>();
        for (String line : ledgerLines(content)) uris.add(Uri.parse(line));
        return uris;
    }

    private static String readLedger(Context ctx) {
        File file = new File(ctx.getFilesDir(), LEDGER);
        if (!file.exists()) return "";
        try (InputStream in = new java.io.FileInputStream(file)) {
            return new String(readAll(in), StandardCharsets.UTF_8);
        } catch (IOException e) {
            Log.w(TAG, "could not read the intake ledger: " + e.getMessage());
            return "";
        }
    }

    /** True when the ledger on disk now says what it was asked to say. */
    private static boolean writeLedger(Context ctx, String content) {
        File file = new File(ctx.getFilesDir(), LEDGER);
        if (content.isEmpty()) {
            if (file.exists() && !file.delete()) {
                Log.w(TAG, "could not clear the intake ledger");
                return false;
            }
            return true;
        }
        // Same rename-into-place as store(): a half-written ledger would either lose a
        // pending route or resurrect a delivered one.
        File part = new File(ctx.getCacheDir(), LEDGER + ".part");
        try (FileOutputStream out = new FileOutputStream(part)) {
            out.write(content.getBytes(StandardCharsets.UTF_8));
        } catch (IOException e) {
            Log.w(TAG, "could not write the intake ledger: " + e.getMessage());
            if (!part.delete()) Log.w(TAG, "could not delete " + part.getName());
            return false;
        }
        if (!part.renameTo(file)) {
            Log.w(TAG, "could not move the intake ledger into place");
            if (!part.delete()) Log.w(TAG, "could not delete " + part.getName());
            return false;
        }
        return true;
    }

    /* The three below are pure so they can be tested without a Context. */

    static List<String> ledgerLines(String content) {
        List<String> lines = new ArrayList<>();
        for (String line : content.split("\n")) {
            String trimmed = line.trim();
            if (!trimmed.isEmpty()) lines.add(trimmed);
        }
        return lines;
    }

    /** Appends what is not already there: the same URI must not queue up twice. */
    static String ledgerWith(String content, List<String> add) {
        List<String> lines = ledgerLines(content);
        for (String line : add) {
            if (line != null && !line.trim().isEmpty() && !lines.contains(line)) lines.add(line);
        }
        return join(lines);
    }

    static String ledgerWithout(String content, String remove) {
        List<String> lines = ledgerLines(content);
        lines.remove(remove);
        return join(lines);
    }

    private static String join(List<String> lines) {
        if (lines.isEmpty()) return "";
        StringBuilder out = new StringBuilder();
        for (String line : lines) out.append(line).append('\n');
        return out.toString();
    }

    // ---------------------------------------------------------------- reading

    static int pendingCount(Context ctx) {
        File[] files = prune(inbox(ctx).listFiles());
        return files == null ? 0 : files.length;
    }

    /** A route shared while the web layer never got to run would sit here forever. */
    private static File[] prune(File[] files) {
        if (files == null) return null;
        long cutoff = System.currentTimeMillis() - MAX_AGE_MS;
        for (File file : files) {
            if (file.lastModified() < cutoff && !file.delete()) Log.w(TAG, "could not prune " + file.getName());
        }
        return Arrays.stream(files).filter(File::exists).toArray(File[]::new);
    }

    /** Returns the oldest pending route and removes it from the inbox. */
    static Pending next(Context ctx) {
        File[] files = prune(inbox(ctx).listFiles());
        if (files == null || files.length == 0) return null;
        Arrays.sort(files);

        for (File file : files) {
            byte[] data = null;
            try (InputStream in = new java.io.FileInputStream(file)) {
                data = readAll(in);
            } catch (IOException e) {
                Log.w(TAG, "could not read pending route: " + e.getMessage());
            }
            if (!file.delete()) Log.w(TAG, "could not delete " + file.getName());
            if (data == null || data.length == 0) continue;
            return new Pending(displayNameOf(file), decode(data));
        }
        return null;
    }

    // ---------------------------------------------------------------- helpers

    private static File inbox(Context ctx) {
        File dir = new File(ctx.getFilesDir(), DIR);
        if (!dir.exists() && !dir.mkdirs()) Log.w(TAG, "could not create " + dir);
        return dir;
    }

    private static byte[] readAll(InputStream in) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] buf = new byte[8192];
        int read;
        while ((read = in.read(buf)) != -1) {
            out.write(buf, 0, read);
            if (out.size() > MAX_BYTES) throw new IOException("route larger than " + MAX_BYTES + " bytes");
        }
        return out.toByteArray();
    }

    private static String displayName(Context ctx, Uri uri) {
        String name = null;
        if ("content".equals(uri.getScheme())) {
            try (Cursor c = ctx.getContentResolver().query(uri, null, null, null, null)) {
                if (c != null && c.moveToFirst()) {
                    int i = c.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                    if (i >= 0) name = c.getString(i);
                }
            } catch (Exception e) {
                Log.w(TAG, "could not resolve display name: " + e.getMessage());
            }
        }
        if (name == null) name = uri.getLastPathSegment();
        return name == null ? "route.gpx" : name;
    }

    /** Strips the timestamp prefix added by {@link #store}. */
    private static String displayNameOf(File file) {
        String name = file.getName();
        int sep = name.indexOf("__");
        return sep < 0 ? name : name.substring(sep + 2);
    }

    private static boolean hasAllowedExtension(String name) {
        if (name == null) return false;
        int dot = name.lastIndexOf('.');
        return dot >= 0 && ALLOWED_EXTENSIONS.contains(name.substring(dot + 1).toLowerCase());
    }

    /** GPX is XML and normally UTF-8; some exporters still emit Latin-1. */
    static String decode(byte[] data) {
        try {
            return StandardCharsets.UTF_8.newDecoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT)
                    .decode(ByteBuffer.wrap(data))
                    .toString();
        } catch (CharacterCodingException e) {
            // Not valid UTF-8 after all.
            return new String(data, StandardCharsets.ISO_8859_1);
        }
    }

    /** Timestamp+sequence file name: unique across shares, sorts in arrival order. */
    static String inboxFileName(long millis, int sequence, String name) {
        return String.format(Locale.ROOT, "%013d-%04d__%s", millis, sequence % 10000, sanitize(name));
    }

    private static boolean looksLikeRoute(byte[] data) {
        int head = Math.min(data.length, 2048);
        String start = new String(data, 0, head, StandardCharsets.UTF_8).toLowerCase();
        return start.contains("<gpx") || start.contains("<kml");
    }

    private static String sanitize(String name) {
        String base = (name == null || name.isEmpty()) ? "route.gpx" : name;
        base = base.replaceAll("[^A-Za-z0-9._ -]", "-");
        if (!hasAllowedExtension(base)) base = base + ".gpx";
        return base.length() > 120 ? base.substring(base.length() - 120) : base;
    }
}
